R/utilities-simulation-results.R

```diff
@@ -83,6 +83,77 @@ getOutputValues <- function(simulationResults, quantitiesOrPaths, individualIds
   return(output)
 }
 
+#' Export values that can be plugged directly into the TLF lib (TODO)
+#'
+#' @export
+getOutputValuesTLF <- function(simulationResults, population, quantitiesOrPaths = NULL, individualIds = NULL) {
+  validateIsOfType(simulationResults, SimulationResults)
+  validateIsOfType(population, Population)
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+  validateIsOfType(quantitiesOrPaths, c(Quantity, "character"), nullAllowed = TRUE)
+
+  quantitiesOrPaths <- quantitiesOrPaths %||% simulationResults$allQuantityPaths
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(paths, Quantity)) {
+    paths <- unlist(lapply(paths, function(x) x$consolidatePath))
+  }
+  paths <- unique(paths)
+
+  result <- list()
+
+  if (length(paths) == 0) {
+    return(result)
+  }
+
+  # If no specific individual ids are passed, iterate through all individuals
+  individualIds <- ifNotNull(individualIds, unique(individualIds), simulationResults$allIndividualIds)
+
+  # All time values are equal
+  timeValues <- simulationResults$timeValues
+  valueLength <- length(timeValues)
+  covariateNames <- population$allCovariateNames
+
+  metaData <- list(
+    Time = list(unit = "min", dimension = "Time")
+  )
+
+  individualPropertiesCache  <- vector("list", length(individualIds))
+  # create a cache of all indivdual values that are constant independent from the path
+  for (individualIndex in seq_along(individualIds)) {
+    individualId <- individualIds[individualIndex]
+    covariates <- population$covariatesAt(individualId)
+    individualProperties <- list(IndividualId = rep(individualId, valueLength), Time = timeValues)
+
+    for (covariateName in covariateNames) {
+      individualProperties[[covariateName]] <- rep(covariates$valueFor(covariateName), valueLength)
+    }
+
+    individualPropertiesCache[[individualIndex]] <-individualProperties;
+  }
+
+  for (path in paths) {
+    quantity <- getQuantity(path, simulationResults$simulation)
+    resultsForPath <- vector("list", length(individualIds))
+    metaData$Value <- list(unit = quantity$unit, dimension = quantity$dimension)
+    pathColumn <- rep(path, valueLength)
+
+    for (individualIndex in seq_along(individualIds)) {
+      individualId <- individualIds[individualIndex]
+      individualProperties <- individualPropertiesCache[[individualIndex]]
+      values <- simulationResults$getValuesForIndividual(path, individualId)
+      resultsForPath[[individualIndex]] <- data.frame(individualProperties, Path = pathColumn, Value = values)

```

Comment:
Removing the individualProperties from the data.frame removes about 1/3 of the time. 
Lol there is clearly a lot that can be optimized here.